splice_machine/src/main/java/com/splicemachine/derby/stream/control/ControlTableChecker.java

```diff
@@ -134,41 +136,100 @@ private void populateData(PairDataSet table, PairDataSet index) {
+            Tuple2<String, Tuple2<byte[], ExecRow>> t = indexSource.next();
             String baseRowId = t._1;
-            byte[] indexKey = t._2;
-            indexData.put(baseRowId, indexKey);
+            Tuple2<byte[], ExecRow> row = t._2;
+            indexData.put(baseRowId, row);
         }
     }
 
     private List<String> checkDuplicateIndexes() throws StandardException {
-        List<String> messages = new LinkedList<>();
-        ArrayListMultimap<String, byte[]> result = ArrayListMultimap.create();
+        ArrayListMultimap<String, Tuple2<byte[], ExecRow>> result = ArrayListMultimap.create();
         long duplicateIndexCount = 0;
         for (String baseRowId : indexData.keySet()) {
-            List<byte[]> rows = indexData.get(baseRowId);
+            List<Tuple2<byte[], ExecRow>> rows = indexData.get(baseRowId);
             if (rows.size() > 1) {
-                duplicateIndexCount += rows.size();
+                duplicateIndexCount += rows.size()-1;
                 result.putAll(baseRowId, rows);
             }
         }
-        int i = 0;
         if (duplicateIndexCount > 0) {
-            messages.add(String.format("The following %d indexes are duplicates:", duplicateIndexCount));
+            if (fix) {
+                fixDuplicateIndexes(result);
+            }
+            return reportDuplicateIndexes(result, duplicateIndexCount);
+        }
+        return new LinkedList<>();
+    }
+
+    private void fixDuplicateIndexes(ArrayListMultimap<String, Tuple2<byte[], ExecRow>> result) throws StandardException {
+        try {
+            WriteCoordinator writeCoordinator = PipelineDriver.driver().writeCoordinator();
+            WriteConfiguration writeConfiguration = writeCoordinator.defaultWriteConfiguration();
+            Partition indexPartition = SIDriver.driver().getTableFactory().getTable(Long.toString(conglomerate));
+            RecordingCallBuffer<KVPair> writeBuffer = writeCoordinator.writeBuffer(indexPartition, txn, null, writeConfiguration);
+
+            List<Integer> indexToMain = tentativeIndex.getIndex().getIndexColsToMainColMapList();
             for (String baseRowId : result.keySet()) {
-                List<byte[]> indexKeys = result.get(baseRowId);
-                for (byte[] indexKey : indexKeys) {
-                    if (i >= maxCheckTableErrors) {
+                ExecRow baseRow = tableData.get(baseRowId);
+                List<Tuple2<byte[], ExecRow>> indexRows = result.get(baseRowId);
+                for (Tuple2<byte[], ExecRow> indexRow : indexRows) {
+                    boolean duplicate = false;
+                    DataValueDescriptor[] dvds = indexRow._2.getRowArray();
+                    for (int i = 0; i < dvds.length - 1; ++i) {
+                        int col = baseColumnMap[indexToMain.get(i) - 1];
+                        if (!dvds[i].equals(baseRow.getColumn(col + 1))) {
+                            duplicate = true;
+                            break;
```

Comment:
We seem to assume that if there are duplicate index rows, the extra ones must have different values than the base row. Is it possible that the extra index row also has the same values as the base row?

Reply:
The index that has some column values as base row is the correct one and should be kept.